### PR TITLE
feat(air): add boundary constraint to hasher

### PR DESCRIPTION
**Pull request by tohrnii**
Addresses #233.

### CODE REVIEW

File: air/src/aux_table/mod.rs
Comment (grjte) [inline air/src/aux_table/mod.rs:4]: nit: I would prefer to import this from `super`

File: air/src/aux_table/hasher/mod.rs
Comment (grjte) [inline air/src/aux_table/hasher/mod.rs:11]: nit: I would prefer to import both of these from `super`. (It looks like `TransitionConstraintDegree` was auto-imported and overlooked previously.)

File: air/src/aux_table/mod.rs
Comment (grjte) [inline air/src/aux_table/mod.rs:81]: I would prefer to use `get_assertions_first_step` here instead of `get_assertions` so that all of the internal modules follow the same pattern and the only `get_assertions` call is at the top level (in the AIR lib). I think this will keep things more consistent if we need to add any "last step" assertions within the aux table in the future.

File: air/src/lib.rs
Comment (grjte) [inline air/src/lib.rs:150]: Two notes here:
1. As mentioned above, I would prefer to use `aux_table::get_assertions_first_step` instead.
2. I would prefer to move this up so it happens before we start setting assertions for the last step (above the "set assertions for the last step" divider comment)

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @tohrnii! What you've done looks exactly right. I just left a few small comments inline.

**Comment by tohrnii** [comment]
@grjte - Thank you for the review. Made suggested changes. Please let me know if there's anything else that needs to be done to complete this PR.
Also, the hasher is referred to as `hash co-processor` and `hasher co-processor` in different places in comments. Should we make that consistent?

File: docs/src/design/aux_table/hasher.md
Comment (grjte) [inline docs/src/design/aux_table/hasher.md:241]: Based on this comment, there's at least one change required earlier in the documentation as well. I searched back and found a reference at row 39 that should be updated, but there may be others as well. Can you update at row 39 and check the full document again to see if there are any other spots that need to change?

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @tohrnii the code looks great! I think the documentation needs to be double-checked before this can be merged. I left a comment inline.

Regarding your question about the "hash co-processor" vs "hasher co-processor" - yes, I think this is a good point and it would be better for us to be consistent. I'm going to defer to @bobbinth for his preference and we can stick to that going forward.

File: docs/src/design/aux_table/hasher.md
Comment (tohrnii) [reply docs/src/design/aux_table/hasher.md:241]: Thanks. Fixed. Found one other instance in processor::hasher in the comments.

**Comment by tohrnii** [comment]
> I have a slight preference for "hash co-processor" - just because it has fewer letters

Thanks. Changed all instances of `hasher co-processor` to `hash co-processor`

**Review by grjte — APPROVED** [review]
Looks good to me! Thanks @tohrnii!

**Pull request merged**
